probe_src/performance_test.py

```diff
@@ -1,190 +1,224 @@
+        io_counters_raw = process.io_counters()
+        io_counters = {
+            'read_count': io_counters_raw.read_count,
+            'write_count': io_counters_raw.write_count,
+            'read_bytes': io_counters_raw.read_bytes,
+            'write_bytes': io_counters_raw.write_bytes
+        }
+    except Exception as e:
+        print(f"Failed to get I/O counters: {e}")
+
+    return cpu_times, memory_usage, io_counters
+
+def benchmark_command(command: str, warmup_iterations: int, benchmark_iterations: int, flag: bool) -> list[Result]:
     results = []
 
     for _ in range(warmup_iterations):
+        print(f"Running warmup command: {command}")
+        subprocess.run("rm probe_log", shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
         subprocess.run(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
 
-    for _ in range(benchmark_iterations):
-        start_time_psutil = datetime.datetime.now()
-
-        proc = psutil.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-
-        cpu_times = memory_info = io_counters = None
-
-        if proc.is_running():
-            try:
-                cpu_times = proc.cpu_times()
-                memory_info = proc.memory_info()
-                io_counters = proc.io_counters()
-            except (psutil.NoSuchProcess, psutil.AccessDenied):
-                pass
-
-        stdout, stderr = proc.communicate()
-
-        datetime.datetime.now()
-
-        datetime.datetime.now()
-        returncode = proc.wait()
-        end_time_wait = datetime.datetime.now()
-
-        result = Result(
-            returncode=returncode,
-            cpu_times=cpu_times,
-            memory_info=memory_info,
-            io_counters=io_counters,
-            stdout=stdout.decode('utf-8'),
-            stderr=stderr.decode('utf-8'),
-            start_time=start_time_psutil,
-            end_time=end_time_wait
-        )
-        results.append(result)
-        time.sleep(4)  # Pause for 4 seconds after each command
+    if not flag:
+        for _ in range(benchmark_iterations):
+            cleanup()
+            start_time_psutil = datetime.datetime.now()
+            print(f"Starting process with command: {command}")
+            proc = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            stdout, stderr = proc.communicate()
+            print(f"Command STDOUT:\n{stdout.decode('utf-8')}")
+            print(f"Command STDERR:\n{stderr.decode('utf-8')}")
+            returncode = proc.wait()
+            end_time_wait = datetime.datetime.now()
+            cpu_times, memory_usage, io_counters = get_process_stats(proc)
+            total_cpu_time = cpu_times[0] + cpu_times[1]
+            elapsed_time = (end_time_wait - start_time_psutil).total_seconds()
+            cpu_usage_percent = (total_cpu_time / elapsed_time) * 100 if elapsed_time > 0 else 0
+            result = Result(
```

Comment:
We should include all of these lines in `get_process_stats`, which could probably just return a `Result` object, since the process of computing that is the same for every case (just the command differs).